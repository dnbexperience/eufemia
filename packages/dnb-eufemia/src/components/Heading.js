/**
 * ATTENTION: This file is auto generated by using "prepareTemplates".
 * Do not change the content!
 *
 */

/**
 * Library Index heading to autogenerate all the components and extensions
 * Used by "prepareHeadings"
 */

import Heading from './heading/Heading'
export * from './heading/Heading'
export default Heading
