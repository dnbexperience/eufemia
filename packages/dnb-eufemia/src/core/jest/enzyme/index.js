/**
 * Default Component Test Setup
 *
 */

// enzyme setup
const enzyme = require('enzyme')

// TODO: This may not be needed in future
const Adapter = require('enzyme-adapter-react-16')
enzyme.configure({ adapter: new Adapter() })

module.exports = enzyme
