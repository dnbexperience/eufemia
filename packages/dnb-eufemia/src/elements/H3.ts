/**
 * This file is created manually
 * You can change the content
 *
 */

import H3 from './typography/H3'
export * from './typography/H3'
export default H3
