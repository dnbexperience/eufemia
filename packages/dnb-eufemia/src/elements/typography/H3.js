/**
 * HTML Element
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import { spacingPropTypes } from '../components/space/SpacingHelper'
import H from './H'

const H3 = (props) => <H as="h3" {...props} />
H3.propTypes = {
  ...spacingPropTypes,
  level: PropTypes.string,
  size: PropTypes.oneOf([
    'xx-large',
    'x-large',
    'large',
    'medium',
    'basis',
    'small',
    'x-small',
  ]),
  children: PropTypes.node,
}
H3.defaultProps = {
  level: null,
  size: 'medium',
  children: null,
}

export default H3
