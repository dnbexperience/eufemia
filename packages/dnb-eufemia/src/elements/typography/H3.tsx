/**
 * HTML Element
 *
 */
import React from 'react'
import type { SharedHProps } from './H'
import H from './H'

const H3 = ({ size, ...props }: SharedHProps) => (
  <H as="h3" size={size || 'auto'} {...props} />
)

H3._isHeadingElement = true
H._supportsSpacingProps = true

export default H3
