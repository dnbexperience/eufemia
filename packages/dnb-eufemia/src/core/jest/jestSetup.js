/**
 * Jest Setup
 *
 */

import { axe, toHaveNoViolations } from 'jest-axe'
import fakeProps from 'react-fake-props'
import { shallow, mount, render } from './enzyme'
import ReactDOMServer from 'react-dom/server'
import fs from 'fs-extra'
import onceImporter from 'node-sass-once-importer'
import path from 'path'
import sass from 'node-sass'
import { toBeType } from 'jest-tobetype'
import toJson from 'enzyme-to-json'

export {
  fakeProps, // we have also our own replacement function called "fakeAllProps"
  shallow,
  mount,
  render,
  toJson,
  axe,
  toHaveNoViolations
}

global.IS_TEST = true
if (typeof window !== 'undefined') {
  window.IS_TEST = true
}

expect.extend({ toBeType })
expect.extend(toHaveNoViolations)

export const loadScss = (file, options = {}) => {
  try {
    const sassResult = sass.renderSync({
      file,
      includePaths: [path.resolve(__dirname, '../../style/core/')],
      importer: [onceImporter()],
      ...options
    })
    return String(sassResult.css)
  } catch (e) {
    console.error('loadScss error:', e)
    return e
  }
}

export const mockGetSelection = () => {
  let memory
  Object.defineProperty(window.navigator, 'clipboard', {
    configurable: true,
    value: {
      writeText: jest.fn().mockImplementation((v) => {
        memory = v
        return Promise.resolve(v)
      }),
      readText: jest.fn().mockImplementation(() => Promise.resolve(memory))
    }
  })

  const mockRange = new (class Range {
    constructor() {
      this.startContainer = {
        parentNode: document.createElement('div')
      }
    }
    getElement() {
      return this.node
    }
    insertNode(elem) {
      this.node = document.createElement('div')
      this.node.appendChild(elem)
      return this
    }
    cloneRange() {
      return this
    }
  })()

  let ranges = [mockRange]

  const mockValue = '1234.56'
  let value = mockValue
  let rangeCount = 9

  class RangeObj {
    rangeCount = rangeCount
    toString = () => value
    addRange(range = mockRange) {
      value = mockValue
      ranges.push(range)
      rangeCount = ranges.length
    }
    getRangeAt(index) {
      return ranges[index]
    }
    removeAllRanges() {
      value = ''
      ranges = []
      rangeCount = ranges.length
    }
  }

  Object.defineProperty(window, 'getSelection', {
    configurable: true,
    value: () => {
      return new RangeObj()
    }
  })
}

export const loadImage = async (imagePath) =>
  await fs.readFile(path.resolve(imagePath))

export const toHtml = (Component) =>
  ReactDOMServer.renderToStaticMarkup(Component)

export const axeComponent = async (...components) => {
  const html = components
    .filter((Component) =>
      // enzyme names the mounted wrapper: ReactWrapper
      /react/i.test(String(Component.constructor))
    )
    .map((Component) => toHtml(Component))
    .join('\n')

  return await axe(
    `<main>${html}</main>`,
    typeof components[1] === 'object' ? components[1] : null
  )
}

export function attachToBody() {
  let container = document.createElement('div')
  document.body.append(container)
  return container
}

function jestSetup() {}

export default jestSetup
