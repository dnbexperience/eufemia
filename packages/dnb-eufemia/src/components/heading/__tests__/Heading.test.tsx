/**
 * Component Test
 *
 */

import React from 'react'
import { axeComponent, loadScss } from '../../../core/jest/jestSetup'
import { render } from '@testing-library/react'
import Heading, {
  resetLevels,
  setNextLevel,
  HeadingProps,
  HeadingLevel,
} from '../Heading'
import H3 from '../../../elements/H3'

const warn = jest.fn()

describe('Heading component', () => {
  it('renders with empty props', () => {
    const props: HeadingProps = {}
    render(<Heading {...props} />)

    expect(document.querySelector('.dnb-heading')).toBeInTheDocument()
  })

  it('have to match level correction', () => {
    render(
      <React.StrictMode>
        <Heading level={2} debug={warn}>
          Heading #1
        </Heading>
        <Heading.Level reset={1} debug={warn}>
          <Heading level={2}>Heading #2</Heading>
          <Heading level={3}>Heading #3</Heading>
          <Heading level={4}>Heading #4</Heading>
          <Heading level={4}>Heading #5</Heading>
          <Heading level={2}>Heading #6</Heading>
          <Heading level={1}>Heading #7</Heading>
        </Heading.Level>
        <Heading.Level debug={warn}>
          <Heading>Heading #8</Heading>
        </Heading.Level>
        <Heading.Level increase debug={warn}>
          <Heading>Heading #9</Heading>
        </Heading.Level>
        <Heading.Level inherit debug={warn}>
          <Heading>Heading #10</Heading>
        </Heading.Level>
        <Heading debug inherit>
          Heading #11
        </Heading>
        <Heading debug inherit decrease>
          Heading #12
        </Heading>
      </React.StrictMode>
    )

    let i = -1
    const elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] Heading #1')
    expect(elem[++i].textContent).toBe('[h1] Heading #2')
    expect(elem[++i].textContent).toBe('[h2] Heading #3')
    expect(elem[++i].textContent).toBe('[h3] Heading #4')
    expect(elem[++i].textContent).toBe('[h4] Heading #5')
    expect(elem[++i].textContent).toBe('[h2] Heading #6')
    expect(elem[++i].textContent).toBe('[h2] Heading #7')
    expect(elem[++i].textContent).toBe('[h2] Heading #8')
    expect(elem[++i].textContent).toBe('[h3] Heading #9')
    expect(elem[++i].textContent).toBe('[h3] Heading #10')
    expect(elem[++i].textContent).toBe('[h3] Heading #11')
    expect(elem[++i].textContent).toBe('[h2] Heading #12')
  })

  it('have to match global reset', () => {
    render(
      <React.StrictMode>
        <Heading.Level debug={warn} reset={1}>
          <Heading>Heading #1</Heading>
        </Heading.Level>

        <Heading.Level debug={warn} reset={1}>
          <Heading>Heading #2</Heading>
        </Heading.Level>

        <Heading debug>Heading #3</Heading>
        <Heading debug reset>
          Heading #4
        </Heading>
      </React.StrictMode>
    )

    let i = -1
    const elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] Heading #1')
    expect(elem[++i].textContent).toBe('[h1] Heading #2')
    expect(elem[++i].textContent).toBe('[h2] Heading #3')
    expect(elem[++i].textContent).toBe('[h2] Heading #4')
  })

  it('have to match context reset', () => {
    render(
      <React.StrictMode>
        <Heading.Level debug={warn} reset={1}>
          <Heading>Heading #1</Heading>
          <Heading>Heading #2</Heading>
          <Heading increase>Heading #3</Heading>
          <Heading.Level reset>
            <Heading>Heading #4</Heading>
          </Heading.Level>
          <Heading reset>Heading #5</Heading>
        </Heading.Level>
      </React.StrictMode>
    )

    let i = -1
    const elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] Heading #1')
    expect(elem[++i].textContent).toBe('[h2] Heading #2')
    expect(elem[++i].textContent).toBe('[h3] Heading #3')
    expect(elem[++i].textContent).toBe('[h2] Heading #4')
    expect(elem[++i].textContent).toBe('[h2] Heading #5')
  })

  it('have to match level correction with manual heading', () => {
    render(
      <React.StrictMode>
        <Heading.Level debug={warn} reset={1}>
          <Heading>Heading #1</Heading>
          <Heading>Heading #2</Heading>
          <H3 level="use">Heading #3</H3>
          <Heading>Heading #4</Heading>
        </Heading.Level>
      </React.StrictMode>
    )

    const first = document.querySelectorAll('h3.dnb-h--medium')
    expect(first[0].textContent).toBe('Heading #3')

    let i = -1
    const elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] Heading #1')
    expect(elem[++i].textContent).toBe('[h2] Heading #2')
    expect(elem[++i].textContent).toBe('[h3] Heading #4')
  })

  it('have to match after level state update', () => {
    const warn = jest.fn()

    const RenderComp = (props) => (
      <React.StrictMode>
        <Heading debug={warn} {...props}>
          Heading #1
        </Heading>
      </React.StrictMode>
    )
    const { rerender } = render(<RenderComp />)

    expect(document.querySelector('.dnb-heading').textContent).toBe(
      '[h1] Heading #1'
    )

    rerender(<RenderComp level={3} />)

    // We got a level correction here!
    expect(document.querySelector('.dnb-heading').textContent).toBe(
      '[h2] Heading #1'
    )

    // expect(warn). toHaveBeenCalledTimes(2) // 2 because of StrictMode
    expect(warn).toHaveBeenCalledWith(
      'Heading levels can only be changed by factor one! Got:',
      3,
      'and had before',
      1,
      '- The new level is',
      2,
      '\nNB: This warning was triggered by:',
      '',
      'Heading #1'
    )

    rerender(<RenderComp level={4} skip_correction={true} />)

    expect(document.querySelector('.dnb-heading').textContent).toBe(
      '[h4] Heading #1'
    )
    // still one time, same as we had earlier
    expect(warn).toHaveBeenCalledTimes(1) // 2 because of StrictMode
  })

  it('have to have correct leveling after using setNextLevel', async () => {
    const { rerender: rerenderH1, container: h1 } = render(
      <Heading debug={warn}>h1</Heading>
    )

    Heading.setNextLevel(2)

    const { rerender: rerenderH2, container: h2 } = render(
      <Heading debug={warn}>h2</Heading>
    )

    setNextLevel(3)

    const { rerender: rerenderH3, container: h3 } = render(
      <Heading.Level debug={warn}>
        <Heading>h3</Heading>
      </Heading.Level>
    )

    expect(h1.querySelector('.dnb-heading').textContent).toBe('[h1] h1')
    expect(h2.querySelector('.dnb-heading').textContent).toBe('[h2] h2')
    expect(h3.querySelector('.dnb-heading').textContent).toBe('[h3] h3')

    Heading.setNextLevel(4)

    rerenderH2(
      <Heading debug={warn} level={3}>
        h2
      </Heading>
    )
    expect(h2.querySelector('.dnb-heading').textContent).toBe('[h4] h2')

    setNextLevel(1)

    rerenderH2(
      <Heading debug={warn} level={5}>
        h2
      </Heading>
    )
    expect(h2.querySelector('.dnb-heading').textContent).toBe('[h1] h2')

    rerenderH2(
      <Heading debug={warn} level={2}>
        h2
      </Heading>
    )
    expect(h2.querySelector('.dnb-heading').textContent).toBe('[h2] h2')

    setNextLevel(2)

    rerenderH1(
      <Heading debug={warn} level={3}>
        h1
      </Heading>
    )
    expect(h1.querySelector('.dnb-heading').textContent).toBe('[h2] h1')

    rerenderH3(
      <Heading debug={warn} level={3}>
        h1
      </Heading>
    )
    expect(h3.querySelector('.dnb-heading').textContent).toBe('[h3] h1')
  })

  it('have to have correct leveling after using resetLevels', () => {
    const { rerender, container: h2 } = render(
      <Heading debug={warn}>h2</Heading>
    )
    expect(h2.querySelector('.dnb-heading').textContent).toBe('[h1] h2')

    rerender(
      <Heading debug={warn} level={3}>
        h2
      </Heading>
    )
    expect(h2.querySelector('.dnb-heading').textContent).toBe('[h2] h2')

    resetLevels(1)

    rerender(
      <Heading debug={warn} level={4}>
        h2
      </Heading>
    )
    expect(h2.querySelector('.dnb-heading').textContent).toBe('[h1] h2')

    rerender(
      <Heading debug={warn} level={2}>
        h2
      </Heading>
    )
    expect(h2.querySelector('.dnb-heading').textContent).toBe('[h2] h2')
  })

  it('have to have aria role and level if set as span element', () => {
    render(
      <Heading element="span" debug={warn} reset={1}>
        Heading #1
      </Heading>
    )

    const elem = document.querySelectorAll('span.dnb-heading')
    expect(elem[0].textContent).toBe('[h1] Heading #1')
    expect(elem[0].getAttribute('role')).toBe('heading')
    expect(elem[0].getAttribute('aria-level')).toBe('1')
  })

  it('have to refuse to set level below 1', () => {
    render(
      <Heading debug={warn} level={0 as HeadingLevel} reset={1}>
        Heading #1
      </Heading>
    )

    expect(document.querySelectorAll('.dnb-heading')[0].textContent).toBe(
      '[h1] Heading #1'
    )
  })

  it('have to have correct size class', () => {
    render(
      <Heading debug={warn} size="x-large" reset={1}>
        Heading #1
      </Heading>
    )

    const elem = document.querySelectorAll('.dnb-heading')
    expect(elem[0].textContent).toBe('[h1] Heading #1')
    expect(elem[0].classList).toContain('dnb-h--x-large')
    expect(elem[0].getAttribute('class')).toBe(
      'dnb-heading dnb-h--x-large'
    )
  })

  it('should set level if skip_correction is true', () => {
    render(
      <React.StrictMode>
        <Heading.Level debug={warn} skip_correction reset={1}>
          <Heading level={4}>Heading #1</Heading>
          <Heading increase>Heading #2</Heading>
        </Heading.Level>
      </React.StrictMode>
    )

    const elem = document.querySelectorAll('.dnb-heading')
    expect(elem[0].textContent).toBe('[h4] Heading #1')
    expect(elem[1].textContent).toBe('[h5] Heading #2')
  })

  it('should not increase level above 6', () => {
    render(
      <React.StrictMode>
        <Heading.Level debug={warn}>
          <Heading>Heading #1</Heading>
          <Heading.Increase skip_correction level={6}>
            <Heading>Heading #2</Heading>
            <Heading increase>Heading #3</Heading>
          </Heading.Increase>
        </Heading.Level>
      </React.StrictMode>
    )

    const elem = document.querySelectorAll('.dnb-heading')
    expect(elem[0].textContent).toBe('[h1] Heading #1')
    expect(elem[1].textContent).toBe('[h6] Heading #2')
    expect(elem[2].textContent).toBe('[h6] Heading #3')
  })

  it('should keep context level after state update', () => {
    const { rerender } = render(<StateChanges />)

    let i = -1
    let elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] h1')
    expect(elem[++i].textContent).toBe('[h2] h2')
    expect(elem[++i].textContent).toBe('[h3] h3')
    expect(elem[++i].textContent).toBe('[h3] h3 before')
    expect(elem[++i].textContent).toBe('[h3] h3 after')
    expect(elem[++i].textContent).toBe('[h2] h2')

    rerender(<StateChanges showHeading3={true} />)

    i = -1
    elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] h1')
    expect(elem[++i].textContent).toBe('[h2] h2')
    expect(elem[++i].textContent).toBe('[h3] h3')
    expect(elem[++i].textContent).toBe('[h3] h3 before')
    expect(elem[++i].textContent).toBe('[h4] h4 1')
    expect(elem[++i].textContent).toBe('[h4] h4 2')
    expect(elem[++i].textContent).toBe('[h5] h5 1')
    expect(elem[++i].textContent).toBe('[h3] h3 after')
    expect(elem[++i].textContent).toBe('[h2] h2')

    rerender(<StateChanges showHeading4={true} />)

    i = -1
    elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] h1')
    expect(elem[++i].textContent).toBe('[h2] h2')
    expect(elem[++i].textContent).toBe('[h3] h3')
    expect(elem[++i].textContent).toBe('[h3] h3 before')
    expect(elem[++i].textContent).toBe('[h4] h4 1')
    expect(elem[++i].textContent).toBe('[h4] h4 2')
    expect(elem[++i].textContent).toBe('[h5] h5 1')
    expect(elem[++i].textContent).toBe('[h3] h3 after')
    expect(elem[++i].textContent).toBe('[h4] h4 1')
    expect(elem[++i].textContent).toBe('[h4] h4 2')
    expect(elem[++i].textContent).toBe('[h5] h5 1')
    expect(elem[++i].textContent).toBe('[h2] h2')

    rerender(<StateChanges showHeading4={false} />)

    // also test to reset the context, as this should be truthy
    setNextLevel(1)

    rerender(<StateChanges showHeading4={true} />)

    i = -1
    elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] h1')
    expect(elem[++i].textContent).toBe('[h2] h2')
    expect(elem[++i].textContent).toBe('[h3] h3')
    expect(elem[++i].textContent).toBe('[h3] h3 before')
    expect(elem[++i].textContent).toBe('[h4] h4 1')
    expect(elem[++i].textContent).toBe('[h4] h4 2')
    expect(elem[++i].textContent).toBe('[h5] h5 1')
    expect(elem[++i].textContent).toBe('[h3] h3 after')
    expect(elem[++i].textContent).toBe('[h4] h4 1')
    expect(elem[++i].textContent).toBe('[h4] h4 2')
    expect(elem[++i].textContent).toBe('[h5] h5 1')
    expect(elem[++i].textContent).toBe('[h2] h2')
  })

  it('have to match default leveling', () => {
    render(
      <React.StrictMode>
        <Heading.Level debug={warn} reset={1}>
          <Heading>Heading #1</Heading>
          <Heading>Heading #2</Heading>

          <Heading.Increase>
            <Heading>Heading #3</Heading>
            <Heading up>Heading #4</Heading>
            <Heading down>Heading #5</Heading>
            <Heading decrease>Heading #6</Heading>
          </Heading.Increase>

          <Heading.Increase>
            <Heading>Heading #7</Heading>
            <Heading.Increase decrease>
              <Heading>Heading #8</Heading>
              <Heading decrease>Heading #9</Heading>
            </Heading.Increase>
          </Heading.Increase>
        </Heading.Level>
      </React.StrictMode>
    )

    let i = -1
    const elem = document.querySelectorAll('.dnb-heading')
    expect(elem[++i].textContent).toBe('[h1] Heading #1')
    expect(elem[++i].textContent).toBe('[h2] Heading #2')
    expect(elem[++i].textContent).toBe('[h3] Heading #3')
    expect(elem[++i].textContent).toBe('[h4] Heading #4')
    expect(elem[++i].textContent).toBe('[h3] Heading #5')
    expect(elem[++i].textContent).toBe('[h2] Heading #6')
    expect(elem[++i].textContent).toBe('[h3] Heading #7')
    expect(elem[++i].textContent).toBe('[h2] Heading #8')
    expect(elem[++i].textContent).toBe('[h2] Heading #9')
  })

  it('should validate with ARIA rules', async () => {
    const Comp = render(
      <React.StrictMode>
        <Heading.Level debug={warn} reset={1}>
          <Heading>Heading #1</Heading>
          <Heading>Heading #2</Heading>

          <Heading.Increase>
            <Heading>Heading #3</Heading>
            <Heading up>Heading #4</Heading>
            <Heading down>Heading #5</Heading>
            <Heading decrease>Heading #6</Heading>
          </Heading.Increase>

          <Heading.Increase>
            <Heading>Heading #7</Heading>
            <Heading.Increase decrease>
              <Heading>Heading #8</Heading>
              <Heading decrease>Heading #9</Heading>
            </Heading.Increase>
          </Heading.Increase>
        </Heading.Level>
      </React.StrictMode>
    )
    expect(await axeComponent(Comp, {})).toHaveNoViolations()
  })

  it('has to match style dependencies css', () => {
    const css = loadScss(require.resolve('../style/deps.scss'))
    expect(css).toMatchSnapshot()
  })
})

let gComp
// eslint-disable-next-line no-unused-vars
function makeComp() {
  gComp =
    gComp ||
    render(
      <React.StrictMode>
        <Heading.Level debug={warn} reset={1}>
          <Heading>Heading #1</Heading>
          <Heading>Heading #2</Heading>

          <Heading.Increase>
            <Heading>Heading #3</Heading>
            <Heading up>Heading #4</Heading>
            <Heading down>Heading #5</Heading>
            <Heading decrease>Heading #6</Heading>
          </Heading.Increase>

          <Heading.Increase>
            <Heading>Heading #7</Heading>
            <Heading.Increase decrease>
              <Heading>Heading #8</Heading>
              <Heading decrease>Heading #9</Heading>
            </Heading.Increase>
          </Heading.Increase>
        </Heading.Level>
      </React.StrictMode>
    )

  return gComp
}

function StateChanges({
  showHeading3 = false,
  showHeading4 = false,
} = {}) {
  const [state, setState] = React.useState({
    showHeading3,
    showHeading4,
  })

  React.useEffect(() => {
    setState({ showHeading3, showHeading4: state.showHeading4 })
  }, [showHeading3]) // eslint-disable-line react-hooks/exhaustive-deps

  React.useEffect(() => {
    setState({ showHeading4, showHeading3: state.showHeading3 })
  }, [showHeading4]) // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <Heading.Level group="A" debug={warn} reset={1}>
      <Heading>h1</Heading>
      <Heading>h2</Heading>
      <Heading increase>h3</Heading>

      <Heading.Level group="B">
        <Heading>h3 before</Heading>
        {state.showHeading3 && (
          <React.StrictMode>
            <Heading increase>h4 1</Heading>
            <Heading>h4 2</Heading>
            <Heading increase>h5 1</Heading>
          </React.StrictMode>
        )}
        <Heading>h3 after</Heading>

        <Heading.Increase group="C">
          {state.showHeading4 && (
            <React.StrictMode>
              <Heading>h4 1</Heading>
              <Heading>h4 2</Heading>
              <Heading increase>h5 1</Heading>
            </React.StrictMode>
          )}
        </Heading.Increase>
      </Heading.Level>

      <Heading.Decrease group="C">
        <Heading>h2</Heading>
      </Heading.Decrease>
    </Heading.Level>
  )
}
