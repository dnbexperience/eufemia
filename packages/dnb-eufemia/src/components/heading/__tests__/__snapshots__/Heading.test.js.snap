// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading component have to match default heading snapshot 1`] = `
<HeadingProvider
  counter={null}
  debug={
    [MockFunction] {
      "calls": Array [
        Array [
          "Heading levels can only be changed by factor one! Got:",
          3,
          "and had before",
          1,
          "- The new level is",
          2,
          "
NB: This warning was triggered by:",
          "",
          "Heading #3",
        ],
        Array [
          "Heading levels can only be changed by factor one! Got:",
          4,
          "and had before",
          2,
          "- The new level is",
          3,
          "
NB: This warning was triggered by:",
          "",
          "Heading #4",
        ],
        Array [
          "Can not decrease to heading level 1! Had before",
          2,
          "- The new level is",
          2,
          "
NB: This warning was triggered by:",
          "",
          "Heading #7",
        ],
        Array [
          "Can not [increment] heading level higher than 6! Got:",
          7,
          "and had before",
          6,
          "- The new level is",
          6,
          "
NB: This warning was triggered by:",
          "",
          "Heading #3",
        ],
        Array [
          "Can not decrease to heading level 1! Had before",
          2,
          "- The new level is",
          2,
          "
NB: This warning was triggered by:",
          "",
          "Heading #9",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  decrease={null}
  down={null}
  group={null}
  id={null}
  increase={null}
  inherit={null}
  level={null}
  reset={1}
  text={null}
  up={null}
>
  <Heading
    class={null}
    className={null}
    counter={null}
    debug={null}
    debug_counter={null}
    decrease={null}
    down={null}
    element="auto"
    group={null}
    id={null}
    increase={null}
    inherit={null}
    level={null}
    reset={null}
    size="auto"
    skeleton={null}
    skip_correction={null}
    text={null}
    up={null}
  >
    <h1
      className="dnb-heading dnb-h--xx-large"
    >
      <span
        className="dnb-heading__debug"
      >
        [h1] 
      </span>
      Heading #1
    </h1>
  </Heading>
  <Heading
    class={null}
    className={null}
    counter={null}
    debug={null}
    debug_counter={null}
    decrease={null}
    down={null}
    element="auto"
    group={null}
    id={null}
    increase={null}
    inherit={null}
    level={null}
    reset={null}
    size="auto"
    skeleton={null}
    skip_correction={null}
    text={null}
    up={null}
  >
    <h2
      className="dnb-heading dnb-h--large"
    >
      <span
        className="dnb-heading__debug"
      >
        [h2] 
      </span>
      Heading #2
    </h2>
  </Heading>
  <Component>
    <HeadingProvider
      counter={null}
      decrease={null}
      down={null}
      group={null}
      id={null}
      increase={true}
      inherit={null}
      level={null}
      reset={null}
      text={null}
      up={null}
    >
      <Heading
        class={null}
        className={null}
        counter={null}
        debug={null}
        debug_counter={null}
        decrease={null}
        down={null}
        element="auto"
        group={null}
        id={null}
        increase={null}
        inherit={null}
        level={null}
        reset={null}
        size="auto"
        skeleton={null}
        skip_correction={null}
        text={null}
        up={null}
      >
        <h3
          className="dnb-heading dnb-h--medium"
        >
          <span
            className="dnb-heading__debug"
          >
            [h3] 
          </span>
          Heading #3
        </h3>
      </Heading>
      <Heading
        class={null}
        className={null}
        counter={null}
        debug={null}
        debug_counter={null}
        decrease={null}
        down={null}
        element="auto"
        group={null}
        id={null}
        increase={null}
        inherit={null}
        level={null}
        reset={null}
        size="auto"
        skeleton={null}
        skip_correction={null}
        text={null}
        up={true}
      >
        <h4
          className="dnb-heading dnb-h--basis"
        >
          <span
            className="dnb-heading__debug"
          >
            [h4] 
          </span>
          Heading #4
        </h4>
      </Heading>
      <Heading
        class={null}
        className={null}
        counter={null}
        debug={null}
        debug_counter={null}
        decrease={null}
        down={true}
        element="auto"
        group={null}
        id={null}
        increase={null}
        inherit={null}
        level={null}
        reset={null}
        size="auto"
        skeleton={null}
        skip_correction={null}
        text={null}
        up={null}
      >
        <h3
          className="dnb-heading dnb-h--medium"
        >
          <span
            className="dnb-heading__debug"
          >
            [h3] 
          </span>
          Heading #5
        </h3>
      </Heading>
      <Heading
        class={null}
        className={null}
        counter={null}
        debug={null}
        debug_counter={null}
        decrease={true}
        down={null}
        element="auto"
        group={null}
        id={null}
        increase={null}
        inherit={null}
        level={null}
        reset={null}
        size="auto"
        skeleton={null}
        skip_correction={null}
        text={null}
        up={null}
      >
        <h2
          className="dnb-heading dnb-h--large"
        >
          <span
            className="dnb-heading__debug"
          >
            [h2] 
          </span>
          Heading #6
        </h2>
      </Heading>
    </HeadingProvider>
  </Component>
  <Component>
    <HeadingProvider
      counter={null}
      decrease={null}
      down={null}
      group={null}
      id={null}
      increase={true}
      inherit={null}
      level={null}
      reset={null}
      text={null}
      up={null}
    >
      <Heading
        class={null}
        className={null}
        counter={null}
        debug={null}
        debug_counter={null}
        decrease={null}
        down={null}
        element="auto"
        group={null}
        id={null}
        increase={null}
        inherit={null}
        level={null}
        reset={null}
        size="auto"
        skeleton={null}
        skip_correction={null}
        text={null}
        up={null}
      >
        <h3
          className="dnb-heading dnb-h--medium"
        >
          <span
            className="dnb-heading__debug"
          >
            [h3] 
          </span>
          Heading #7
        </h3>
      </Heading>
      <Component
        decrease={true}
      >
        <HeadingProvider
          counter={null}
          decrease={true}
          down={null}
          group={null}
          id={null}
          increase={true}
          inherit={null}
          level={null}
          reset={null}
          text={null}
          up={null}
        >
          <Heading
            class={null}
            className={null}
            counter={null}
            debug={null}
            debug_counter={null}
            decrease={null}
            down={null}
            element="auto"
            group={null}
            id={null}
            increase={null}
            inherit={null}
            level={null}
            reset={null}
            size="auto"
            skeleton={null}
            skip_correction={null}
            text={null}
            up={null}
          >
            <h2
              className="dnb-heading dnb-h--large"
            >
              <span
                className="dnb-heading__debug"
              >
                [h2] 
              </span>
              Heading #8
            </h2>
          </Heading>
          <Heading
            class={null}
            className={null}
            counter={null}
            debug={null}
            debug_counter={null}
            decrease={true}
            down={null}
            element="auto"
            group={null}
            id={null}
            increase={null}
            inherit={null}
            level={null}
            reset={null}
            size="auto"
            skeleton={null}
            skip_correction={null}
            text={null}
            up={null}
          >
            <h2
              className="dnb-heading dnb-h--large"
            >
              <span
                className="dnb-heading__debug"
              >
                [h2] 
              </span>
              Heading #9
            </h2>
          </Heading>
        </HeadingProvider>
      </Component>
    </HeadingProvider>
  </Component>
</HeadingProvider>
`;

exports[`Heading component have to match snapshot 1`] = `
"/*
* DNB Heading
*
*/
/**
 * This file is only used to make components independent
 * so that they can get imported individually, without the core styles
 *
 */
/*
 * Utilities
 */
/*
 * Scopes
 *
 */
/*
 * Document Reset
 *
 */
.dnb-suffix {
  padding-left: 0.5rem;
  font-size: var(--font-size-basis);
  line-height: var(--line-height-basis);
  word-break: normal; }

/*
 * Heading Component
 *
 */
.dnb-heading__debug {
  color: var(--color-black-30);
  font-weight: var(--font-weight-basis); }
  .dnb-heading__debug .dnb-code {
    color: var(--color-black);
    font-size: 12px !important; }

.dnb-heading__context {
  display: block;
  padding: 1rem 0 2rem; }
"
`;
